README.md
# Minecraft Manager 3

[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0) ![GitHub package.json version (branch)](https://img.shields.io/github/package-json/v/theemeraldtree/minecraft-manager/major)

Minecraft Manager is an experimental Minecraft launcher that lets you easily manage multiple instances. You can install mods, modpacks, and more in a clean, user-friendly interface.

## Features

- Install mods, modpacks, resource packs, and worlds directly from CurseForge
- Modern, user-friendly, and powerful interface
- Supports import and export for Twitch and MultiMC formats 

Not all features are currently working. **This is a development branch. Nothing is guaranteed to be stable**.

## Getting Started

If you'd like prebuilt binaries, [you can find them here for Windows, macOS, and Linux](https://theemeraldtree.net/mcm/download).

If you would like to build it and run it for yourself, it's pretty simple

### Clone the repo & install packages

```
git clone https://github.com/theemeraldtree/minecraft-manager.git
cd minecraft-manager
yarn
```

Yarn should be used because of the lockfile.

### Running in a development environment

```
yarn dev
```

This will spawn the Electron process and start webpack-dev-server on port 9483.

### Building

The simplest way to build Minecraft Manager is to run

```
yarn package
```

This will compile Webpack, then run electron-builder and compile for your current OS and architecture.
However you can also compile Webpack and electron-builder separately.

To build webpack:

```
yarn compile
```

This will compile Webpack with the settings defined in `webpack.config.json` and place the output in /bundles

To run electron-builder

```
yarn build
```

This will run electron-builder for your current OS and architecture, and place the output in /dist

### Credits

Credits are viewable in the Settings menu of Minecraft Manager. You can also find a list of every package used in `package.json`.

### License

Minecraft Manager is licensed under the [GNU General Public License v3.0](https://github.com/theemeraldtree/minecraft-manager/blob/master/LICENSE), a free software license. If you have improvements you would like to make to Minecraft Manager, please consider making a Pull Request - contributions are welcome!


CHANGELOG.md
# Changelog

Changelogs for development versions are located in the work branch

## 3.0.0
**In development**

### Removed
- Launcher Integration
  - The feature was hard to maintain, and I don't think it received much usage.

## 2.6.0

#### Released 2020-08-02

### Features

- Completely new and redesigned UI layout
- Experimental "support" for legacy Forge (it's very buggy)
- New edit version page
- Redesigned editing mods, worlds, resource packs, and datapacks

### Changes

- "Profiles" have been renamed to "Instances"
- Improved CurseForge search
- Analytics have been removed
- Import no longer requires you to specify type, it's now inferred

### Fixes

- Forge fixes
- Curse fixes

## 2.5.1

#### Released 2020-06-25

### Fixes

- Fixed a bug where Forge on 1.13+ would fail to launch; you will need to reinstall Forge

## 2.5.0

#### Released 2020-06-24

### Features

- Built-in Java Management
- Redesigned Settings page
- An error message will now occur if app initialization fails
- Improved asset editing, allowing you to drag-and-drop mods, resourcepacks, and worlds
- New welcome page
- Twitch ZIP format import
- You can now manually add Mojang accounts from within the launcher
- Mojang account access tokens will automatically be refreshed
- Minecraft crash reports will be shown if the game crashes
- Custom Java arguments and a custom Java install can now be used
- Profiles can now override certain global settings
- New post-update changelog

### Changes

- Launcher Integration is now optional
- Internal changes to modloaders with new patch files
- A profile's respective assets will now update if changes to the folder are made
- Improved asset importing
- Reduced app bundle size
- Each game instance is now run in a seperate bin folder
- Downloading of Minecraft game assets is now significantly faster
- Internal changes to Version JSON files
- Improved failed to launch error handling
- Imrproved startup time by using v8-compile-cache

### Fixes

- "Select file" dialogs will work now
- Network errors won't fail silently
- Welcome page won't faily silently
- Profile deletion won't fail silently
- Memory leaks with asynchronous network requests have been fixed
- "Too long" errors when launching a profile should no longer occur
- Forge 1.13+ should work better now
- You will now be warned if you don't have an account
- Non-directories in the profiles folder will no longer throw errors

## 2.4.7

#### Released 2020-05-25

### Fixes

- Linux now starts up correctly

## 2.4.6

#### Released 2020-05-25

### Fixes

- Critical problem in the Welcome page

## 2.4.5

#### Released 2020-05-24

### Fixes

- Profiles displayed on the profiles screen had an extra "a" at the end

## 2.4.4

#### Released 2020-05-23

### Features

- Massive performance improvements

### Fixes

- Search in Discover is no longer slow

## 2.4.3

#### Released 2020-05-18

### Features

- New, experimental privacy-respecting analytics
- New hover effect on buttons and cards

### Changed

- Small UI improvements

## 2.4.2

#### Released 2020-05-11

### Changed

- The application is now installed per user instead of per system

### Fixed

- Crash loading mods list
- Update throwing error
- Fixed bugs in Welcome page
- Missing mods from Curse cancelling install

### Removed

- Removed .mcjprofile file association

## 2.4.1

#### Released 2020-05-09

### Features

- New sidebar design in Edit Pages and Settings
- Improved UI consistency
- Better usage via keyboard
- Improved Error Handling

## 2.4.0

#### Released 2020-04-13

### Features

- Added a cool loading spinner
- Launch directly (bypass the Minecraft Launcher)
- Better error handling
- Minecraft Options, OptiFine Options, and Servers list are automatically copied on Profile creation
- Downloads Viewer is now transparent

### Removed

- Removed "in-game defaults" (e.g. autojump, tutorial) as those are now inferred from your Latest release Options

## 2.3.8

#### Released 2020-04-10

### Security

- Fixed "kind-of" security vulnerability (CVE-2019-20149)

### Features

- Detailed logging is now saved to disk

### Fixes

- Fixed inconsistencies in capitalization

## 2.3.7

#### Released 2020-04-08

### Fixes

- Fixed Dependencies not downloading correctly
- Fixed Fabric not showing install complete
- Fixed Snapshots running in seperate folders
- Fixed "Open Profile Folder" button

### Features

- Allow syncing of Minecraft Options, OptiFine Options, and Servers
- Allow copying of Options and Servers

### Changed

- Reorganized Edit Advanced page

## 2.3.6

#### Released 2020-03-23

### Fixes

- Fixed Downloads not appearing in the viewer
- Fixed a small Toast bug

## 2.3.5

#### Released 2020-03-22

### Features

- Dropdowns for Minecraft Versions now include search and type selection
- Sort and Filter when adding assets from CurseForge
- Improve modloader installation
  - Allow installing of custom versions
- Add Full License Disclaimer to About Page
- Show Downloads info in Asset Cards
- The option to disable News checks on start up
- MultiMC Importing (experimental)
- Release date is now shown on versions
- Added the option to run the Snapshot profile in the same game directory as Latest release

### Changed

- Adjusted spacing between Profiles
- "Framework is installing" persists between edit page changes
- Downloads are now logged in the main process console
- Improve CurseForge HTML rendering
- Improve About page
- Improve Update dialog

### Fixes

- Fixed Notice Toasts staying on the screen for a long time
- Fix clicking away from dropdowns
- Fixed Minecraft Version selection bugs
- Fix all Toasts dissappearing when one was dismissed

## 2.3.4

#### Released 2020-03-17

Version for testing GitHub Actions and Releases. No new features or bug fixes.

## 2.3.3

#### Released 2020-03-17

Version for testing GitHub Actions and Releases. No new features or bug fixes.

## 2.3.2

#### Released 2020-03-17


main.js
const { app, BrowserWindow, shell, ipcMain, Notification, dialog } = require('electron');
const { autoUpdater } = require('electron-updater');
const url = require('url');
const path = require('path');
const fs = require('fs');
const axios = require('axios');
const os = require('os');
const winston = require('winston');
const { execFile } = require('child_process');
require('winston-daily-rotate-file');
require('v8-compile-cache');

const singleInstanceLock = app.requestSingleInstanceLock();
if (!singleInstanceLock) app.quit();

const logformat = winston.format.printf(
  ({ timestamp, level, message }) => `[${timestamp}] [${level.toUpperCase()}]: ${message}`
);

const rotateFile = new winston.transports.DailyRotateFile({
  filename: '%DATE%-main.log',
  datePattern: 'YYYY-MM-DD',
  maxFiles: '7d',
  createSymlink: true,
  symlinkName: 'latest-main.log',
  auditFile: path.join(app.getPath('userData'), '/logs/main-node-process/.mcm-main-log-audit.json'),
  dirname: path.join(app.getPath('userData'), '/logs/main-node-process')
});

const mainLogger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(winston.format.timestamp(), winston.format.align(), logformat),
  transports: [rotateFile, new winston.transports.Console({ colorize: true, prettyPrint: true, level: 'info' })]
});

mainLogger.info('Launching Minecraft Manager');
mainLogger.info('Starting Main Node Process');

// Security warning IS DISABLED because we're loading from localhost.
// this is only disabled so it doesn't clog the console in dev mode
process.env.ELECTRON_DISABLE_SECURITY_WARNINGS = true;

ipcMain.on('install-update', () => {
  autoUpdater.quitAndInstall();
});

const downloadProgresses = {};

ipcMain.on('download-file', async (event, downloadURL, dest, id) => {
  mainLogger.info('Received Download File IPC');
  downloadProgresses[id] = 0;
  let progressData = 0;
  const ws = fs.createWriteStream(dest);
  mainLogger.info(`Starting download of ${downloadURL}`);
  try {
    const { data, headers } = await axios(downloadURL, {
      responseType: 'stream',
      headers: {
        'X-Client': 'MinecraftManager'
      }
    });

    const contentLength = headers['content-length'];

    data.on('data', chunk => {
      progressData += chunk.length;

      const prog = Math.trunc((progressData / contentLength) * 100);
      if (prog - downloadProgresses[id] >= 10) {
        downloadProgresses[id] = prog;
        if (event) {
          if (event.sender) {
            event.sender.send('file-download-progress', {
              id,
              progress: prog
            });
          }
        }
      }
    });

    data.pipe(ws);

    ws.on('finish', () => {
      mainLogger.info(`Finished download of ${downloadURL}`);
      event.sender.send('file-download-finish', {
        id
      });
    });
  } catch (e) {
    mainLogger.error(e);
    event.sender.send('file-download-error', { id, error: e });

    ws.close();

    if (fs.existsSync(dest)) {
      fs.unlinkSync(dest);
    }
  }
});

const dev = require('process').execPath.includes('electron');

let mainWindow;

ipcMain.on('start-progress', () => {
  if (mainWindow) {
    mainWindow.setProgressBar(2);
  }
});

ipcMain.on('stop-progress', () => {
  if (mainWindow) {
    mainWindow.setProgressBar(0);
  }
});

function navigation(event, navURL) {
  if (
    navURL.substring(0, 22) !== 'http://localhost:9483/' ||
    navURL.substring(21, 29) === '/linkout' ||
    navURL.substring(32, 40) === '/linkout'
  ) {
    let finalUrl = navURL;

    // Handle Curse's "You're leaving CurseForge" feature
    // we redirect straight to the page
    if (navURL.substring(21, 29) === '/linkout') {
      // Curse encodes the URI's twice for some reason
      finalUrl = decodeURIComponent(decodeURI(navURL.substring(40)));
    }

    // linkout for minecraft.curseforge.com
    if (navURL.substring(32, 40) === '/linkout') {
      finalUrl = navURL.substring(51);
    }

    mainLogger.info(`Navigating externally to ${finalUrl}`);

    event.preventDefault();
    shell.openExternal(finalUrl);
  }
}

function createWindow() {
  // why show a frame on linux but hide it on everything else?
  // on linux CSD (client-side decoration) causes a ton of problems without a frame
  // so unfortunately it has to be disabled
  mainLogger.info('Creating Electron BrowserWindow');

  mainWindow = new BrowserWindow({
    width: 800,
    height: 800,
    frame: os.platform() === 'linux',
    backgroundColor: '#222',
    webPreferences: {
      webSecurity: false,
      experimentalFeatures: true,
      nodeIntegration: true
    },
    titleBarStyle: 'hidden'
  });

  mainWindow.removeMenu();

  if (os.platform() === 'win32') {
    if (fs.existsSync(path.join('C:\\Program Files\\Minecraft Manager\\Uninstall Minecraft Manager.exe'))) {
      // 2.4.2 changed the program from being a per-machine install to a per-user install
      // Unfortunately electron-updater doesn't cleanup the old version,
      // so we have to check and remove it manually
      const ret = dialog.showMessageBox({
        type: 'info',
        title: 'Heads Up!',
        message:
          'Minecraft Manager version 2.4.2 requires the uninstaller of the previous version to be ran on first launch.\n\nYou may have to agree to some User Account Control dialogs.\n\nIf you get an error, report it at https://theemeraldtree.net/mcm/issues',
        buttons: ['View More Info and Continue', 'Continue'],
        noLink: true,
        defaultId: 1,
        cancelId: 1
      });

      if (ret === 0) {
        shell.openExternal('https://github.com/theemeraldtree/minecraft-manager/wiki/2.4.2-Uninstaller-Information');
      }

      execFile('C:\\Program Files\\Minecraft Manager\\Uninstall Minecraft Manager.exe', ['/S']);
    }
  }

  mainWindow.ELECTRON_DISABLE_SECURITY_WARNINGS = true;

  if (dev) {
    // install the react devtools
    // make sure you create a .reactDevtools.json file that looks similar to the following
    // {
    //    "extPath": "C:\\Users\\username\\AppData\\Local\\Google\\Chrome\\User Data\\\Default\\Extensions\\extensionidgoeshere\\versionnumber"
    // }

    if (fs.existsSync('.reactDevtools.json')) {
      mainLogger.info('Attempting to add React DevTools to BrowserWindow');
      const json = JSON.parse(fs.readFileSync('.reactDevtools.json'));
      BrowserWindow.addDevToolsExtension(path.join(json.extPath));
    }

    mainLogger.info('Opening DevTools in BrowserWindow');
    mainWindow.openDevTools();
  }

  let index;
  if (dev && process.argv.indexOf('--noDevServer') === -1) {
    index = url.format({
      protocol: 'http:',
      host: 'localhost:9483',
      pathname: 'index.html',
      slashes: true
    });
  } else {
    index = url.format({
      protocol: 'file:',
      pathname: path.resolve(__dirname, 'bundles', 'index.html'),
      slashes: true
    });
  }
  mainWindow.loadURL(index);

  mainWindow.once('ready-to-show', () => {
    mainWindow.show();
  });

  mainWindow.webContents.on('will-navigate', (event, navURL) => {
    navigation(event, navURL);
  });

  mainWindow.webContents.on('new-window', (event, navURL) => {
    navigation(event, navURL);
  });
}

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    mainLogger.info('Quitting Minecraft Manager...');
    app.quit();
  }
});

let currentUpdateState = 'checking-for-update';

function checkForUpdates() {
  // if (dev) {
  //   mainLogger.info('[Updater] Cannot update while in dev mode');
  //   mainWindow.webContents.send('in-dev');
  // } else {
  currentUpdateState = 'checking-for-update';
  mainLogger.info('[Updater] Checking for updates...');
  autoUpdater.checkForUpdates();
  autoUpdater.on('checking-for-update', () => {
    mainWindow.webContents.send('checking-for-update');
  });

  autoUpdater.on('update-available', () => {
    currentUpdateState = 'update-available';
    mainLogger.info('[Updater] Found an update. Sending IPC...');
    mainWindow.webContents.send('update-available');
  });

  autoUpdater.on('update-downloaded', () => {
    currentUpdateState = 'update-downloaded';
    mainLogger.info('[Updater] Successfully downloaded update. Sending Notification and IPC...');
    const notif = new Notification({
      title: 'Minecraft Manager',
      body:
        'An update is available for Minecraft Manager. It has been downloaded and will be installed next time you start the app.'
    });
    notif.show();
    mainWindow.webContents.send('update-downloaded');
  });

  autoUpdater.on('update-not-available', () => {
    currentUpdateState = 'update-not-available';
    mainLogger.info('[Updater] No updates available');
    mainWindow.webContents.send('update-not-available');
  });

  autoUpdater.on('error', error => {
    currentUpdateState = 'error';
    mainLogger.error(`[Updater] Error checking for updates:\n${error.toString()}`);
    mainWindow.webContents.send('error', error);
  });
  // }
}

ipcMain.on('check-for-updates', () => {
  if (currentUpdateState !== 'error' && currentUpdateState !== 'update-not-available') {
    mainWindow.webContents.send(currentUpdateState);
  } else {
    checkForUpdates();
  }
});

checkForUpdates();

app.on('ready', createWindow);
